Use a heap-backed file buffer only on wasm while preserving native mmap.

--- a/sys-utils/dmesg.c
+++ b/sys-utils/dmesg.c
@@ -694,9 +694,19 @@
 	if ((intmax_t)st.st_size >= (intmax_t)SINT_MAX(ssize_t))
 		err(EXIT_FAILURE, _("%s is too large"), ctl->filename);
 
+#ifdef __wasm__
+	if (st.st_size <= 0) {
+		errno = EINVAL;
+		err(EXIT_FAILURE, _("cannot read: %s"), ctl->filename);
+	}
+	*buf = xmalloc(st.st_size);
+	if (read_all(fd, *buf, st.st_size) != st.st_size)
+		err(EXIT_FAILURE, _("cannot read: %s"), ctl->filename);
+#else
 	*buf = mmap(NULL, st.st_size, PROT_READ, MAP_SHARED, fd, 0);
 	if (*buf == MAP_FAILED)
 		err(EXIT_FAILURE, _("cannot mmap: %s"), ctl->filename);
+#endif
 	ctl->mmap_buff = *buf;
 	ctl->pagesize = getpagesize();
 	close(fd);
@@ -768,8 +778,12 @@
 
 static void release_buffer(struct dmesg_control *ctl, char *buf)
 {
+#ifdef __wasm__
+	free(buf);
+#else
 	if (!ctl->mmap_buff)
 		free(buf);
+#endif
 	if (ctl->kmsg >= 0)
 		close(ctl->kmsg);
 	if (ctl->json && ul_jsonwrt_is_ready(&ctl->jfmt)) {
@@ -915,12 +929,14 @@
 	/*
 	 * Unmap already printed file data from memory
 	 */
+#ifndef __wasm__
 	if (ctl->mmap_buff && (size_t) (rec->next - ctl->mmap_buff) > ctl->pagesize) {
 		void *x = ctl->mmap_buff;
 
 		ctl->mmap_buff += ctl->pagesize;
 		munmap(x, ctl->pagesize);
 	}
+#endif
 
 	for (i = 0; i < rec->next_size; i++) {
 		const char *p = rec->next + i;
@@ -1035,13 +1051,19 @@
 		 */
 		while (size && !ferror(stdout)) {
 			size_t sz = size > ctl->pagesize ? ctl->pagesize : size;
+#ifdef __wasm__
+			safe_fwrite(ctl, buf, sz, 0, stdout);
+			lastc = buf[sz - 1];
+			buf += sz;
+#else
 			char *x = ctl->mmap_buff;
 
 			safe_fwrite(ctl, x, sz, 0, stdout);
 			lastc = x[sz - 1];
-			size -= sz;
 			ctl->mmap_buff += sz;
 			munmap(x, sz);
+#endif
+			size -= sz;
 		}
 	}
 
